packages/proxy/lib/http/util/service-worker-injector.ts

```diff
@@ -0,0 +1,234 @@
+/// <reference lib="dom" />
+
+import type { ServiceWorkerClientEvent } from './service-worker-manager'
+
+// this should be of type ServiceWorkerGlobalScope, but we can't reference it here
+// because this file is imported elsewhere and the ServiceWorker lib causes
+// errors in those other packages
+declare let self: typeof globalThis & {
+  __cypressServiceWorkerClientEvent: (event: string) => void
+  onfetch: (event: Event) => void
+}
+
+/**
+ * Injects code into the service worker to overwrite the fetch events to determine if the service worker handled the request.
+ * @param body the body of the service worker
+ * @returns the updated service worker
+ */
+export const injectIntoServiceWorker = (body: Buffer) => {
+  function __cypressInjectIntoServiceWorker () {
+    let listenerCount = 0
+    let eventQueue: ServiceWorkerClientEvent[] = []
+    const nonCaptureListenersMap = new WeakMap<EventListenerOrEventListenerObject, EventListenerOrEventListenerObject>()
+    const captureListenersMap = new WeakMap<EventListenerOrEventListenerObject, EventListenerOrEventListenerObject>()
+    const targetToWrappedHandleEventMap = new WeakMap<Object, EventListenerOrEventListenerObject>()
+    const targetToOrigHandleEventMap = new WeakMap<Object, EventListenerOrEventListenerObject>()
+
+    const sendEvent = (event: ServiceWorkerClientEvent) => {
+      // if the binding has been created, we can call it
+      // otherwise, we need to queue the event
+      if (self.__cypressServiceWorkerClientEvent) {
+        self.__cypressServiceWorkerClientEvent(JSON.stringify(event))
+      } else {
+        eventQueue.push(event)
+      }
+    }
+
+    const sendHasFetchEventHandlers = () => {
+      // @ts-expect-error __cypressIsScriptEvaluated is declared below
+      // if the script has been evaluated, we can call the CDP binding to inform the backend whether or not the service worker has a handler
+      if (__cypressIsScriptEvaluated) {
+        sendEvent({ type: 'hasFetchHandler', payload: { hasFetchHandler: !!(listenerCount > 0 || self.onfetch) } })
+      }
+    }
+
+    // A listener is considered valid if it is a function or an object (with the handleEvent function or the function could be added later)
+    const isValidListener = (listener: EventListenerOrEventListenerObject) => {
+      return listener && (typeof listener === 'function' || typeof listener === 'object')
+    }
+
+    // Determine if the event listener was aborted
+    const isAborted = (options?: boolean | AddEventListenerOptions) => {
+      return typeof options === 'object' && options.signal?.aborted
+    }
+
+    // Get the capture value from the options
+    const getCaptureValue = (options?: boolean | AddEventListenerOptions) => {
+      return typeof options === 'boolean' ? options : options?.capture
+    }
+
+    function wrapListener (listener: Function) {
+      return (event) => {
```

Comment:
Updated [18629cf](https://github.com/cypress-io/cypress/pull/28900/commits/18629cf33c74a34158bec26773e9a08cc7fee1dd). I could not directly reference the `webworker` lib since it causes typescipt errors in other packages.